### PR TITLE
ENH: column-wise DataFrame.fillna with Series/Dict value

### LINKED ISSUE #4514: column-wise fillna with Series/dict NotImplemented
As per discussion on [this SO question](http://stackoverflow.com/questions/18127160/pandas-fillna-output-still-has-nan-values) is NotImplementedError.

Solution/workaround is to transpose do transpose? This is used elsewhere in DataFrame.fillna method. just raise if inplace?

cc @cpcloud 

```
In [9]: df = pd.DataFrame([[np.nan, np.nan], [np.nan, 4], [5, 6]], columns=list('AB'))

In [10]: df
Out[10]:
    A   B
0 NaN NaN
1 NaN   4
2   5   6

In [11]: df.mean(0)
Out[11]:
A    5
B    5
dtype: float64

In [12]: df.fillna(df.mean())
Out[12]:
   A  B
0  5  5
1  5  4
2  5  6

In [13]: df.mean(1)
Out[13]:
0    NaN
1    4.0
2    5.5
dtype: float64

In [14]: df.fillna(df.mean(1), axis=1)
---------------------------------------------------------------------------
NotImplementedError                       Traceback (most recent call last)
<ipython-input-14-aecc493431e2> in <module>()
----> 1 df.fillna(df.mean(1), axis=1)

/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/pandas/core/frame.pyc in fillna(self, value, method, axis, inplace, limit, downcast)
   3452             if isinstance(value, (dict, Series)):
   3453                 if axis == 1:
-> 3454                     raise NotImplementedError('Currently only can fill '
   3455                                               'with dict/Series column '
   3456                                               'by column')

NotImplementedError: Currently only can fill with dict/Series column by column
```

### PR DESCRIPTION
**Pull request by arw2019**
- [x] closes #4514
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Picking up #30922

### CODE REVIEW

File: doc/source/whatsnew/v1.2.0.rst [hunk 1]
```diff
@@ -260,6 +260,7 @@ Other enhancements
 - Added :meth:`~DataFrame.set_flags` for setting table-wide flags on a Series or DataFrame (:issue:`28394`)
 - :meth:`DataFrame.applymap` now supports ``na_action`` (:issue:`23803`)
 - :class:`Index` with object dtype supports division and multiplication (:issue:`34160`)
+- :meth:`DataFrame.fillna` can fill NA values column-wise with a dictionary or :class:`Series` (:issue:`4514`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.2.0.rst:263]: move to 1.3

File: pandas/tests/frame/methods/test_fillna.py [hunk 2]
```diff
@@ -525,6 +521,97 @@ def test_fill_corner(self, float_frame, float_string_frame):
...
+        df = DataFrame([[np.nan, np.nan], [np.nan, 4], [5, 6]], columns=list("AB"))
+        result = df.fillna(fill_value, axis=1)
+        tm.assert_frame_equal(expected, result)
+
+    def test_fillna_column_wise_downcast(self):
+        # GH 4514
+        df = DataFrame([[np.nan, 2], [3, np.nan], [np.nan, np.nan]], columns=list("AB"))
+        s = Series([100, 200, 300])
+
+        expected = DataFrame(
```
Comment (jreback) [inline pandas/tests/frame/methods/test_fillna.py:554]: what happens if we have a datetime column mixed in here?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/frame/methods/test_fillna.py (hunk 2 above)
Comment (arw2019) [reply pandas/tests/frame/methods/test_fillna.py:554]: We end up with object dtype:
``` python
In [5]: df = pd.DataFrame([[np.nan, 2], [3, np.nan], [np.nan, np.nan]], columns=list("AB"))
   ...: s = pd.Series(pd.to_datetime([100, 200, 300], unit="ns"))
   ...: 
   ...: result = df.fillna(s, axis=1, downcast="infer")
   ...: result
Out[5]: 
                               A                              B
0  1970-01-01 00:00:00.000000100                            2.0
1                            3.0  1970-01-01 00:00:00.000000200
2  1970-01-01 00:00:00.000000300  1970-01-01 00:00:00.000000300

In [6]: result.dtypes
Out[6]: 
A    object
B    object
dtype: object
```
``` python
In [8]: df = pd.DataFrame({"A": pd.to_datetime([np.nan, 2, np.nan]), "B": pd.to_datetime([3, np.nan, np.nan])})
   ...: s = pd.Series([100, 200, 300])
   ...: 
   ...: result = df.fillna(s, axis=1, downcast="infer")
   ...: result
Out[8]: 
                               A                              B
0                            100  1970-01-01 00:00:00.000000003
1  1970-01-01 00:00:00.000000002                            200
2                            300                            300

In [9]: result.dtypes
Out[9]: 
A    object
B    object
```

File: doc/source/whatsnew/v1.2.0.rst (hunk 1 above)
Comment (arw2019) [reply doc/source/whatsnew/v1.2.0.rst]: Done

File: pandas/tests/frame/methods/test_fillna.py (hunk 2 above)
Comment (arw2019) [reply pandas/tests/frame/methods/test_fillna.py:554]: is this what we want?

**Comment by arw2019** [comment]
Green + addressed comments

**Comment by mroeschke** [comment]
Thanks for the PR but it appears this has gone stale. Looks fairly close but let us know if you'd like to continue by merging master and targeting 1.4

**Pull request closed without merging**